pkg/crmutil/controller-data.go

```diff
@@ -544,20 +613,64 @@ func (cd *CRMHandler) appInstInfoRuntime(ctx context.Context, sender edgeproto.A
 	})
 }
 
+func (cd *CRMHandler) CloudletDNSChanged(ctx context.Context, pf platform.Platform, oldDNS string, new *edgeproto.Cloudlet, sender edgeproto.CloudletInfoSender) (reterr error) {
+	var err error
+
+	fmap := edgeproto.MakeFieldMap(new.Fields)
+	log.SpanLog(ctx, log.DebugLevelInfra, "CloudletDNSChanged", "key", new.Key, "old dns", oldDNS, "new dns", new.RootLbFqdn)
+
+	if !fmap.Has(edgeproto.CloudletFieldRootLbFqdn) {
+		log.SpanLog(ctx, log.DebugLevelApi, "Nothing changed")
+	} else {
+		updateCloudletCallback := sender.SendStatusIgnoreErr
+		err = sender.SendState(dme.CloudletState_CLOUDLET_STATE_UPGRADE, edgeproto.WithSenderResetStatus())
+		if err != nil {
+			return err
+		}
+
+		// TODO -this doesn't send anything back for some reason...
+		updateCloudletCallback(edgeproto.UpdateTask, fmt.Sprintf("3. Updating Cloudlet DNS in CRM - %s", new.Key.Name))
+		err = pf.ChangeCloudletDNS(ctx, new, oldDNS, updateCloudletCallback)
+		if err != nil {
+			err := fmt.Errorf("update failed: %s", err)
+			sender.SendState(dme.CloudletState(edgeproto.TrackedState_UPDATE_ERROR), edgeproto.WithStateError(err))
+			return err
+		}
+	}
+	sender.SendUpdate(func(info *edgeproto.CloudletInfo) error {
+		info.Fields = []string{
+			edgeproto.CloudletInfoFieldState,
+			edgeproto.CloudletInfoFieldStatus,
+		}
+		info.State = dme.CloudletState_CLOUDLET_STATE_READY
+		info.Status.StatusReset()
+		return nil
+	})
+	return nil
+}
+
 func (cd *CRMHandler) CloudletChanged(ctx context.Context, target *edgeproto.CloudletKey, new *edgeproto.Cloudlet, sender edgeproto.CloudletInfoSender) (reterr error) {
 	if new.State == edgeproto.TrackedState_CREATE_REQUESTED || new.State == edgeproto.TrackedState_DELETE_REQUESTED {
 		// This message is for CCRM
 		log.SpanLog(ctx, log.DebugLevelInfra, "cloudletChanged ignoring CCRM state", "cloudlet", new)
 		return
 	}
+
 	// do request
 	log.SpanLog(ctx, log.DebugLevelInfra, "cloudletChanged", "cloudlet", new)
 	pf, err := cd.getPlatform(ctx, target)
 	if err != nil {
 		return err
 	}
-	fmap := edgeproto.MakeFieldMap(new.Fields)
 
+	// Special case for dns update
+	oldDNS, ok := new.Annotations[cloudcommon.AnnotationPreviousDNSName]
+	if ok && oldDNS != new.RootLbFqdn {
+		_ = cd.CloudletDNSChanged(ctx, pf, oldDNS, new, sender)
+		return
+	}
```

Comment:
This should check field flag instead. The field flag is computed from the old state in the cache vs the new received state, so accurately detects a change. The oldDNS annotation is always going to be different from the current RootLbFqdn when this function receives another unrelated update at a later time.

Reply:
Same as above - I keep the check that Annotations actually exist, otherwise this gets triggered on the create.